covergen/covergen.go

```diff
@@ -158,15 +168,122 @@ func check(ct kobo.CoverType, kp, iid string) (string, bool, error) {
 }
 
 func extract(epub string) (image.Image, error) {
-	return nil, errors.New("not implemented")
+	zr, err := zip.OpenReader(epub)
+	if err != nil {
+		return nil, err
+	}
+	defer zr.Close()
+
+	var rootfile string
+	for _, f := range zr.File {
+		if strings.TrimLeft(strings.ToLower(f.Name), "/") == "meta-inf/container.xml" {
+			doc := etree.NewDocument()
+			if rc, err := f.Open(); err != nil {
+				return nil, fmt.Errorf("could not open container.xml: %w", err)
+			} else if _, err = doc.ReadFrom(rc); err != nil {
+				rc.Close()
+				return nil, fmt.Errorf("could not parse container.xml: %w", err)
+			} else {
+				rc.Close()
+			}
+
+			if el := doc.FindElement("//rootfiles/rootfile[@full-path]"); el != nil {
+				rootfile = el.SelectAttrValue("full-path", "")
+			}
+			break
+		}
+	}
+	if rootfile == "" {
+		return nil, errors.New("could not open ebook: could not find package document")
+	}
+
+	var covers []string
+	for _, f := range zr.File {
+		if strings.TrimLeft(strings.ToLower(f.Name), "/") == strings.TrimLeft(strings.ToLower(rootfile), "/") {
+			doc := etree.NewDocument()
+			if rc, err := f.Open(); err != nil {
+				return nil, fmt.Errorf("could not open container.xml: %w", err)
```

Comment:
Looks like a copy paste error here?

Should be something like "could not open package document"?